### ISSUE TITLE
Backfill process too slow

### ISSUE THREAD

**Issue by bsharpe** [rethinkdb/rethinkdb#6055]
We added a new identical node to our cluster of 4 machines this weekend.   

Started sharding tables one at a time -- biggest first (as we've learned; they need the most scratch space).

One table of small rows has 180 millions records in it; it's taken more than 2 days to backfill the shards.  We stopped all other work on the DB to give it the best chance of finishing over the weekend...which it did not.

Either manually (via a setting) or automatically -- seems like the backfill process should utilize more resources when it can.  We would love a MAX setting -- where we could take the DB offline, add a new node; set backfill to MAX and have things resharded in <24hrs.

During the height of the backfill process - I only ever see about 2.5 to 4k writes per second.   At that rate it seems like it should take approximately 13 hrs to copy to the whole table, but ours has taken >50hrs...and it's still going.

Ubuntu 14.04.01 / 256G RAM / 24 dual-core cpus / SSD drives / bonded 10G/s network cxns
using Rethink 2.3.4 with GC tweaks sent to us by Daniel.

**Comment by bsharpe** [comment]
@danielmewes in the meantime... are there any "best practices" for adding nodes to a cluster to minimize read-only access to tables?  

mainly -- we add nodes in order to have more storage space across all nodes -- but in order for that to happen, rethink must re-shard and re-replicate the data which causes a lot of backfilling to happen.

is there a way to seed new nodes with a set of data from one of the other nodes such that it doesn't have to backfill quite so much?

**Comment by danielmewes** [comment]
@bsharpe Hmm hard to say. One thing that could help a bit is if you can move over existing shards without adding any new shards. In that case, downtime should be very very short.

You could take this to the extreme, and shard tables into more shards than necessary once (let's say 16). This can be done through the `reconfigure` command, but not through the web UI (which limits the number of shards to the number of servers in the cluster). After that, you can change the replica assignment of those shards by updating the `table_config` document without adding or removing any shards.
However some queries become less efficient when there are more shards than servers, so this is not without drawbacks.

**Comment by bsharpe** [comment]
@danielmewes Thx.  Although, it's the splitting into shards that takes the time.  We'll take the hit during the "creating more shards than nodes" phase anyway, correct?

We're also wondering if we had an enterprise support contract with you guys would there be anything else that could be done?  That is to say, would it make any difference to us finding a viable solution to growing our cluster?  

This is a significant enough hurdle that we may have to find an alternative datastorage solution as a week without write access to our tables is a deal-breaker for our customers.

**Comment by danielmewes** [comment]
@bsharpe Correct, the potential benefit would be that you would take the hit only once. I'm skeptical to be honest if this is worth it though, since there's also the loss of efficiency due to the over-sharding (this basically will affect any query that uses secondary indexes).

More importantly: I didn't realize you're literally experiencing a week without write access! That is insane. During resharding, write availability should only be impaired for a couple seconds (maybe tens of seconds) periodically throughout the process, but tables should be available for the majority of time. The backfilling itself should not impair availability at all.

The tables will not be "fully available" during this time, but most write operations should be able to go through.

Are you actually getting consistent write failures during the backfilling process?

**Comment by bsharpe** [comment]
@danielmewes TBH, it's been over a month since we added a new node.   

We're currently doing ~150-250k reads/s and ~3-4k writes/s almost all the time.  The last time we added a node was specifically because we were running out of space 2 of our 4 nodes.  We stopped the writes to the db to add the new node, and at the time assumed that things would backfill faster if we weren't trying to write as much to the db as normal.  So, for 2 days we left the writing off.  On the 3rd day, we turned our metric calculations back on because we couldn't go any longer with them off.  The sharding/replicating took another 3 days after that.

Perhaps we were too cautious to start writing heavily to tables that were marked "ready (some replicas missing)" -- but we were concerned about our lack of disk space and eagerly needing the sharding to move data off of the crowded nodes and onto the new one.

So, in the future...  can we assume that as long as a table is marked "ready" (even without all replicas) that we can write to it as normal?

**Comment by danielmewes** [comment]
Yes you can write to it generally. It might indeed slow down the backfill process, if there are concurrent writes. So I guess it's a trade-off between having the resharding finish (slightly) more quickly, and keeping writes running while it's happening.

**Comment by bsharpe** [comment]
@danielmewes Thanks for your time. We love rethink :) and would hate to have to move.
